src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp

```diff
@@ -432,187 +867,194 @@ class npc_memory : public CreatureScript
+            _events.ScheduleEvent(EVENT_WAKING_NIGHTMARE, urand(10 * IN_MILLISECONDS, 13 * IN_MILLISECONDS));
         }
 
         void Initialize()
         {
-            uiOldWoundsTimer = 12000;
-            uiShadowPastTimer = 5000;
-            uiWakingNightmare = 7000;
-        }
+            DoCastAOE(SPELL_SHADOWFORM);
+            DoCastAOE(SPELL_SPAWN_VISUAL);
 
-        uint32 uiOldWoundsTimer;
-        uint32 uiShadowPastTimer;
-        uint32 uiWakingNightmare;
+        }
 
         void Reset() override
         {
+            _events.Reset();
             Initialize();
         }
 
-        void UpdateAI(uint32 uiDiff) override
+
+        void JustEngagedWith(Unit* /*who*/) override
         {
-            if (!UpdateVictim())
-                return;
+            DoZoneInCombat();
+        }
 
-            if (uiOldWoundsTimer <= uiDiff)
-            {
-                if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0))
-                {
-                    if (target->IsAlive())
-                        DoCast(target, SPELL_OLD_WOUNDS);
-                }
-                uiOldWoundsTimer = 12000;
-            }else uiOldWoundsTimer -= uiDiff;
+        void UpdateAI(uint32 diff) override
+        {
+            _events.Update(diff);
 
-            if (uiWakingNightmare <= uiDiff)
-            {
-                DoCast(me, SPELL_WAKING_NIGHTMARE);
-                uiWakingNightmare = 7000;
-            }else uiWakingNightmare -= uiDiff;
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+                return;
 
-            if (uiShadowPastTimer <= uiDiff)
+            while (uint32 eventId = _events.ExecuteEvent())
             {
-                if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 1))
+                switch (eventId)
                 {
-                    if (target->IsAlive())
+                case EVENT_OLD_WOUNDS:
+                    if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 5.0f, true))
+                        DoCast(target, SPELL_OLD_WOUNDS);
+                    _events.Repeat(11 * IN_MILLISECONDS, 13 * IN_MILLISECONDS);
+                    break;
+                case EVENT_SHADOWS_PAST:
+                    if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 40.0f, true))
                         DoCast(target, SPELL_SHADOWS_PAST);
+                    _events.Repeat(5 * IN_MILLISECONDS, 7 * IN_MILLISECONDS);
+                    break;
+                case EVENT_WAKING_NIGHTMARE:
+                    Talk(EMOTE_WAKING_NIGHTMARE);
+                    DoCastAOE(SPELL_WAKING_NIGHTMARE);
+                    _events.Repeat(20 * IN_MILLISECONDS, 40 * IN_MILLISECONDS);
+                    break;
+                default:
+                    break;
                 }
-                uiShadowPastTimer = 5000;
-            }else uiShadowPastTimer -= uiDiff;
+            }
+
+            if (!UpdateVictim())
```

Comment:
why is this after the event handling loop? what event needs to be handled even when not engaged?